### PR TITLE
Task3

### PR DESCRIPTION
**Pull request by Laziriti**
pull requset front app - [https://github.com/Laziriti/aws-rs-school/pull/2](https://github.com/Laziriti/aws-rs-school/pull/2)

link on the front app - [d3bxwsns48af1n.cloudfront.net](d3bxwsns48af1n.cloudfront.net)

Api - [https://ksasg882n4.execute-api.eu-west-1.amazonaws.com/](https://ksasg882n4.execute-api.eu-west-1.amazonaws.com/)

first endpoint - [https://ksasg882n4.execute-api.eu-west-1.amazonaws.com/dev/products/](https://ksasg882n4.execute-api.eu-west-1.amazonaws.com/dev/products/)
respone example:
[{
  "count": 1,
  "description": "string",
  "id": "1",
  "price": 10,
  "title": "string"
}]

second endpoint - [https://ksasg882n4.execute-api.eu-west-1.amazonaws.com/dev/products/{id}](https://ksasg882n4.execute-api.eu-west-1.amazonaws.com/dev/products/{id}) (ids from 1 to 8)
response example:
{
  "count": 1,
  "description": "string",
  "id": "1",
  "price": 10,
  "title": "string"
}

**Additional (optional) tasks (but nice to have):**

Async/await is used in lambda functions **(Done)**
ES6 modules are used for product-service implementation **(Done)**
Webpack is configured for product-service **(Done)**
SWAGGER documentation is created for product-service _(Tried to import my swagger file into api gateway Documentation but there is always error  Invalid Swagger 2.0 input.)_
Lambda handlers are covered by basic UNIT tests (NO infrastructure logic is needed to be covered) **(Done)**
Lambda handlers (getProductsList, getProductsById) code is written not in 1 single module (file) and separated in codebase. **(Done)**
Main error scenarious are handled by API ("Product not found" error, try catch blocks are used in lambda handlers). **(Done)**

### CODE REVIEW

File: .idea/aws-rs-school-node.iml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (DrDiman) [inline .idea/aws-rs-school-node.iml:1]: should these files be here?)

File: product-service-ts/.serverless/cloudformation-template-create-stack.json [hunk 2]
```diff
@@ -0,0 +1,82 @@
+{
```
Comment (DrDiman) [inline product-service-ts/.serverless/cloudformation-template-create-stack.json:1]: the same `.serverless` files should be ignored

File: product-service-ts/package.json [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  "scripts": {
+    "test": "jest --no-cache"
+  },
+  "engines": {
+    "node": ">=14.15.0"
+  },
+  "dependencies": {
+    "@middy/core": "^1.5.2",
+    "@middy/http-json-body-parser": "^1.5.2",
+    "source-map-support": "^0.5.19"
```
Comment (DrDiman) [inline product-service-ts/package.json:15]: is it dependency or devDependency?

File: product-service-ts/src/functions/get-product/handler.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      return formatJSONResponse400({
+        error: 'Bad request, id should be number'
+      });
+    }
+    if (product) {
+      return formatJSONResponse({
+        product
+      });
+    } else {
+      return formatJSONResponse500({
```
Comment (DrDiman) [inline product-service-ts/src/functions/get-product/handler.ts:22]: guess it's 404)

File: product-service-ts/src/tests/getAllPRoducts.test.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        logStreamName: null,
+        getRemainingTimeInMillis: null,
+        done: null,
+        fail: null,
+        succeed: null
+    };
+    it('Should return list of products', async () => {
+        const func = products;
+        let res = await func(eventMock, contestMock, () => { });
+        console.log(res)
```
Comment (DrDiman) [inline product-service-ts/src/tests/getAllPRoducts.test.ts:36]: ?)

File: product-service-ts/src/tests/getAllPRoducts.test.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        logGroupName: null,
+        logStreamName: null,
+        getRemainingTimeInMillis: null,
+        done: null,
+        fail: null,
+        succeed: null
+    };
+    it('Should return list of products', async () => {
+        const func = products;
+        let res = await func(eventMock, contestMock, () => { });
```
Comment (DrDiman) [inline product-service-ts/src/tests/getAllPRoducts.test.ts:35]: `const`s everywhere for sure)

File: product-service-ts/src/tests/getAllPRoducts.test.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        awsRequestId: null,
+        logGroupName: null,
+        logStreamName: null,
+        getRemainingTimeInMillis: null,
+        done: null,
+        fail: null,
+        succeed: null
+    };
+    it('Should return list of products', async () => {
+        const func = products;
```
Comment (DrDiman) [inline product-service-ts/src/tests/getAllPRoducts.test.ts:34]: what is it for? this ref?

File: product-service/handler.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+
+module.exports.hello = async (event) => {
```
Comment (DrDiman) [inline product-service/handler.js:3]: hi)

**Review by DrDiman — COMMENTED** [review]
looks good. 10 points (no swagger documentation)

File: product-service-ts/src/functions/get-all-products/mock.json [hunk 9]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (noway36) [inline product-service-ts/src/functions/get-all-products/mock.json:1]: Do you really need that file?

File: product-service-ts/src/functions/get-all-products/handler.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+import { listOfProducts } from '../../assets/products';
+import schema from './schema';
+
+export const products: ValidatedEventAPIGatewayProxyEvent<typeof schema> = async () => {
+  try {
+    const result = await getAllProduct();
+    return formatJSONResponse(result);
+  } catch (err) {
+    return formatJSONResponse500({
+      error: 'Some error on the server side'
```
Comment (noway36) [inline product-service-ts/src/functions/get-all-products/handler.ts:15]: I guess something like `Filed to get products` would be more descriptive

File: product-service-ts/src/libs/apiGateway.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+export const formatJSONResponse = (response: Record<any, unknown> | Array<unknown>) => {
+  return {
+    statusCode: 200,
+    headers: corsHeaders,
+    body: JSON.stringify(response)
+  };
+};
+
+export const formatJSONResponse500 = (response: Record<string, unknown>) => {
```
Comment (noway36) [inline product-service-ts/src/libs/apiGateway.ts:16]: You could just create one `error response` formatter and pass `statusCode` there

File: product-service-ts/src/functions/get-product/handler.ts [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+import type { ValidatedEventAPIGatewayProxyEvent } from '@libs/apiGateway';
+import { formatJSONResponse, formatJSONResponse500, formatJSONResponse400 } from '../../libs/apiGateway';
+import { middyfy } from '../../libs/lambda';
+import { listOfProducts } from '../../assets/products';
+import schema from './schema';
+
+export const product: ValidatedEventAPIGatewayProxyEvent<typeof schema> = async (event) => {
+  try {
+    const product = await getProductById(event.pathParameters.id);
+    if (!Number(event.pathParameters.id)) {
```
Comment (noway36) [inline product-service-ts/src/functions/get-product/handler.ts:12]: I geuess it is better to validate `id` before invoking `getProductById`

File: product-service-ts/src/functions/get-product/handler.ts [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+        product
+      });
+    } else {
+      return formatJSONResponse500({
+        error: 'There is no product with such id'
+      });
+    }
+  } catch (err) {
+    return formatJSONResponse500({
+      error: 'Some error on the server side'
```
Comment (noway36) [inline product-service-ts/src/functions/get-product/handler.ts:28]: `Failed to get product by id`

File: product-service-ts/src/tests/getAllPRoducts.test.ts [hunk 14]
```diff
@@ -0,0 +1,39 @@
+import { products } from '../functions/get-all-products/handler';
+import { listOfProducts } from '../assets/products';
+
+describe('Get all products', () => {
+    let eventMock = {
```
Comment (noway36) [inline product-service-ts/src/tests/getAllPRoducts.test.ts:5]: `const` and I think you can leave just empty object. To solve the problem with event type you can cast it to `APIGatewayProxyEvent`. (event as APIGatewayProxyEvent)

File: product-service-ts/package.json (hunk 3 above)
Comment (Laziriti) [reply product-service-ts/package.json:15]: Its dependency, was generated by serverless template.

File: product-service-ts/src/tests/getAllPRoducts.test.ts [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+        httpMethod: 'get',
+        path: null,
+        pathParameters: null,
+        isBase64Encoded: null,
+        queryStringParameters: null,
+        stageVariables: null,
+        requestContext: null,
+        resource: null
+    }
+    let contestMock = {
```
Comment (noway36) [inline product-service-ts/src/tests/getAllPRoducts.test.ts:19]: The same about `const`. Similar solution with type casting (`as any as Context`) would help to handle typings.

File: product-service-ts/src/tests/getProductById.test.ts [hunk 16]
```diff
@@ -0,0 +1,82 @@
+import { product } from '../functions/get-product/handler';
+import { listOfProducts } from '../assets/products';
+
+describe('Get product by ID', () => {
+    let eventMockWithId1 = {
```
Comment (noway36) [inline product-service-ts/src/tests/getProductById.test.ts:5]: The same about mock data here and below

File: product-service/serverless.yml [hunk 17]
```diff
@@ -0,0 +1,117 @@
+# Welcome to Serverless!
```
Comment (noway36) [inline product-service/serverless.yml:1]: Do you really need `serverless.yml`? As I see `serverless.ts` file is used to configure sls.

File: product-service-ts/src/tests/getAllPRoducts.test.ts (hunk 15 above)
Comment (noway36) [reply product-service-ts/src/tests/getAllPRoducts.test.ts:19]: Also you can use `aws-lambda-mock-context` if you want

**Comment by noway36** [comment]
LGTM 10 points. 🙂

File: product-service-ts/src/libs/apiGateway.ts (hunk 11 above)
Comment (Laziriti) [reply product-service-ts/src/libs/apiGateway.ts:16]: Yes, but my current implementation more comfort for me.

File: .idea/aws-rs-school-node.iml (hunk 1 above)
Comment (Laziriti) [reply .idea/aws-rs-school-node.iml:1]: Some problems with .gitignore)

File: product-service-ts/src/tests/getAllPRoducts.test.ts (hunk 5 above)
Comment (Laziriti) [reply product-service-ts/src/tests/getAllPRoducts.test.ts]: Fixed, thank you.

File: product-service-ts/src/tests/getAllPRoducts.test.ts (hunk 6 above)
Comment (Laziriti) [reply product-service-ts/src/tests/getAllPRoducts.test.ts]: Fixed)

File: product-service-ts/src/tests/getAllPRoducts.test.ts (hunk 7 above)
Comment (Laziriti) [reply product-service-ts/src/tests/getAllPRoducts.test.ts]: Fixed.

File: product-service/handler.js (hunk 8 above)
Comment (Laziriti) [reply product-service/handler.js]: Forgot to delete this folder)

File: product-service-ts/src/functions/get-all-products/mock.json (hunk 9 above)
Comment (Laziriti) [reply product-service-ts/src/functions/get-all-products/mock.json]: Nope, fixed.

File: product-service-ts/src/functions/get-all-products/handler.ts (hunk 10 above)
Comment (Laziriti) [reply product-service-ts/src/functions/get-all-products/handler.ts]: Fixed.

File: product-service-ts/src/functions/get-product/handler.ts (hunk 12 above)
Comment (Laziriti) [reply product-service-ts/src/functions/get-product/handler.ts:11]: Fixed.

File: product-service-ts/src/functions/get-product/handler.ts (hunk 13 above)
Comment (Laziriti) [reply product-service-ts/src/functions/get-product/handler.ts]: Fixed

File: product-service-ts/src/tests/getAllPRoducts.test.ts (hunk 14 above)
Comment (Laziriti) [reply product-service-ts/src/tests/getAllPRoducts.test.ts]: Fixed.

File: product-service-ts/src/tests/getAllPRoducts.test.ts (hunk 15 above)
Comment (Laziriti) [reply product-service-ts/src/tests/getAllPRoducts.test.ts]: Fixed.

File: product-service-ts/src/tests/getProductById.test.ts (hunk 16 above)
Comment (Laziriti) [reply product-service-ts/src/tests/getProductById.test.ts:6]: Fixed.

File: product-service/serverless.yml (hunk 17 above)
Comment (Laziriti) [reply product-service/serverless.yml]: Forgot to delete this folder (with yml)

File: product-service-ts/.serverless/cloudformation-template-create-stack.json (hunk 2 above)
Comment (Laziriti) [reply product-service-ts/.serverless/cloudformation-template-create-stack.json:1]: Some problems with .gitignore)

**Pull request merged**